File: Python/src/day51/Exercise.py
from math import expm1
class Exercise:
    def ex1ImplementPowerFunction(self, x: float, n: int, c: int) -> float:
        if n==0:
            return 1
        ans = self.ex1ImplementPowerFunction(x, abs(n) // 2, c)
        res = ans * ans % c
        if n%2 == 1:
            res = res * x % c
        if res < 0:
            res = res + c
        return res

    def ex4JosephusProblem(self, n: int, k: int) -> int:
        if n == 1:
            return 1
        return (self.ex4JosephusProblem(n - 1, k) + k-1) % n + 1

File: Java/Intermediate/src/main/java/day51/Exercise.java
package day51;

public class Exercise {
    public static int ex1ImplementPowerFunction(int A, int B, int C){
        if(B==0){
            return 1;
        }
        if(B==1){
            return A%C;
        }
        long ans = ex1ImplementPowerFunction(A, B/2, C);
        ans = (ans*ans)%C;
        if(B%2==1){
            ans = (ans*A)%C;
        }
        if(ans<0){
            ans = ans + C;
        }
        return (int)ans;
    }
}
